### ISSUE TITLE
Apache Arrow support, and other questions

### ISSUE THREAD

**Issue by nevi-me** [pacman82/odbc2parquet#3]
Hi Markus, just discovered this through looking at public crates that depend on Parquet.

I'm one of the Arrow commiters, working on the Rust implementation of Arrow and Parquet.
We recently received a PR that does a lot of refactoring on the Parquet crate, and will very likely come as a breaking change in the next 2-4 months (release 3.0).
If you have some time, may you please look at building your crate with the changes at (https://github.com/apache/arrow/pull/8300#issuecomment-706546221), and provide some feedback if possible.

We are busy with Parquet <> Arrow IO on https://github.com/apache/arrow/tree/rust-parquet-arrow-writer, which we're also hoping to be ready by 3.0.
Out of curiousity, how easy/hard would it be to extend `odbc2parquet` to odbc > arrow, with Parquet then effectively coming for 'free', similar to turbodbc?

We are currently lacking on DB access in Arrow, and that potentially keeps the door closed for people who'd like to optimise Db <> File IO in Rust. I think we could benefit from a lot of allocation improvements in Arrow, so having ODBC to Arrow as an option, would help with ODBC <> CSV and other future formats.

Thanks :smiley:

**Comment by pacman82** [comment]
Hello @nevi-me ,

thanks for starring my humble CLI tool 🙇 . I tested with the branch you suggested and did not encounter any errors.

Using arrow was actually my original thought, but last time I checked the parquet crate did not contain an arrow writer.

The creator of turbodbc is actually a good friend and former colleague of mine at my previous employer. I was hoping to make Rust more popular at my old workplace by demonstrating DB <> File IO in Rust. I've since switched employers and I do work full time Rust now, but in a completly different domain. So this is of somewhat lower priority for me, but I am probaly still down for writing an ODBC <> Arrow crate for the fun of it alone. No promise though.

With regards to ODBC <> CSV : <https://crates.io/crates/odbcsv> crate ;-)

**Comment by nevi-me** [comment]
Hi @pacman82, I'm trying to create an `odbc2arrow` from this crate, but not make it a binary (it works as a binary, but I need a lib).

I've created https://github.com/nevi-me/odbc2arrow, but I'm having `odbc-api` specific issues with the `Environment::new()` not living long enough.

If you have some time, may you please kindlt have a look at it, and advise me on how I can resolve the issues.

I created this so I can try compare with turbodbc for a specific workload that I was trying out today, so this is the first time I look at `odbc-api`. Should I even be using `odbc-api` or something higher-level?

The error I'm getting is at:

```rust
//! file: odbc2arrow.rs [lines 36-63]
let mut odbc_conn = match connection {
    ConnectionType::Dsn(dsn) => {
        // unlike the author of ODBC, we know nothing about the safety here
        // TODO: read up on how this works
        let odbc_env = unsafe { Environment::new() }.unwrap();
        odbc_env.connect(&dsn, user, pass).unwrap()
    }
    ConnectionType::ConnString(conn) => {
        let odbc_env = unsafe { Environment::new() }.unwrap();
        odbc_env.connect_with_connection_string(conn).unwrap()
    }
};

if let Some(cursor) = odbc_conn.exec_direct(query).unwrap() {
    let (schema, buffer_description) = make_arrow_schema(&cursor)?;
    let mut odbc_buffer = OdbcBuffer::new(batch_size, buffer_description.iter().copied());
    let row_set_cursor = cursor.bind_row_set_buffer(&mut odbc_buffer).unwrap();
    Ok(Self {
        schema,
        row_set_cursor,
        batch_size,
        is_complete: false,
    })
} else {
    Err(ArrowError::ComputeError(
        "Query returned no results, unable to determine schema".to_string(),
    ))
}
```

And it's a good ol' lifetime not living long enough

```rust
error[E0597]: `odbc_env` does not live long enough
  --> src\odbc2arrow.rs:40:41
   |
28 | impl<'b, 'o> OdbcArrowReader<'b, 'o> {
   |          -- lifetime `'o` defined here
...
40 |             ConnectionType::Dsn(dsn) => odbc_env.connect(&dsn, user, pass).unwrap(),
   |                                         ^^^^^^^^ borrowed value does not live long enough
...
49 |             let row_set_cursor = cursor.bind_row_set_buffer(&mut odbc_buffer).unwrap();
   |                                  ----------------------------------------------------- assignment requires that `odbc_env` is borrowed for `'o`
...
61 |     }
   |     - `odbc_env` dropped here while still borrowed
```

Thanks

**Comment by pacman82** [comment]
Hello @nevi-me , nice hearing from you. You beat me to the punch, I wanted to start an ODBC <-> Arrow library myself. However, since I am currently still in the process of making input parameters and prepared statements work in `odbc-api`, It would have been at least another month before I would have gotten started. So I am happy about the support in moving the eco-system forward and will gladly lend my advice.

You have actually two issues here:
1. The environment does not live long enough.
2. You MUST NOT wrap `Environment::new` in a safe function call. It can not be done without violating Rusts safety guarantees.

I placed a comment in the safety section of `Environment::new`:

> There may only be one Odbc environment in any process at any time. [...] Also no library should probably wrap the creation of an odbc environment into a safe function call. This is because using two of these "safe" libraries at the same time in different parts of your program may lead to race condition thus violating Rust's safety guarantees.

I agree with making `odbc2arrow` a library. My guesstimate is that you want to provide a safe interface to the user. From that follows strictly that you can not create the ODBC Environment for the user, as only the resulting application can ensure that there is at most one Environment in the entire process. Letting the Application instantiate the Environment for you, will likely fix the lifetime issue as well, as the Environment is no longer owned by the function.

I would actually recommend to narrow the scope even further and not even creating the connection for the user, but start with `odbc_api::Cursor`, which you may also call a result set. My reasoning for this is, that there are lots of ways to acquire a connection (DSN, Connection String, Prompt, usage a pooled connection) and to execute a statement (direct, prepared, with and without parameters). The `odbc-api` is very young and therefore lacks many of these features, but as I mentioned before, at least Parameters and Prepared statements will be added soon. So to not cut the users off, from that functionality (or be forced to forward all that, through your interface), I suppose to start with the things specific to Arrow:

* Converting SQL schema information to Arrow Types
* Binding Buffers to cursor, and if necessary transformation.

Both of which can be implemented on top of a `odbc_api::Cursor`. This should not discourage you from providing convenience functions for common use cases, but exposing something like ODBC Cursor -> Arrow Table directly to the users of the library would make it reusable in many situations, from CLI tool to Web Service.

On another note. Personally I regret naming the tool `odbc2parquet` a little bit since I want to add functionality to also insert the content of parquet files into SQL Tables. I do lack a better name though. Since you might face kind of the same issue with `odbc2arrow`, I would be interested in your thoughts regarding naming, if you are up for some bike-shedding. 

Cheers, Markus

**Comment by nevi-me** [comment]
Thanks Markus,

I got the writing to Arrow files done today (by modifying `odbc2parquet` and using it).
I was able to write a ~4GB files with about 44 million rows, and am able to read it with `pyarrow`.
We had a scenario at work where my colleague noticed that `pyhive`/`pyodbc` were running slow locally compared to the same query in a Spark cluster; so I looked into it.
I replaced pyhive with `turbodbc`, and saw a ~2x speedup, but I was still curious, and decided to try the Rust impl.

In the end, we found that ODBC network was the limitation, as the Rust solution performed similarly to turbodbc.

Thanks for the explanation on the `Environment`, and the recommendation to take in a `Cursor` instead.
It seems like my chief problem is that I might need to hold on to the `RowSetCursor<'_, '_, OdbcBuffer>`.
So even if I take the cursor, I still end up with a lifetime problem. There might be a simple solution, but I don't yet have enough experience with resolving such lifetime issues.

Here's the interface for reference:

```rust
pub struct OdbcArrowReader<'b, 'o> {
    schema: SchemaRef,
    row_set_cursor: RowSetCursor<'b, 'o, OdbcBuffer>,
    batch_size: usize,
    is_complete: bool,
}
```

The convention is to provide an implementation of `arrow::record_batch::RecordBatchReader: Iterator`, so that the user can lazily call the next batch when needed, without being forced to load all data into memory at once.
Do you think this is something that'd be possible, or would we have to consume the whole buffer?

I also bikeshed on https://github.com/nevi-me/rust-dataframe, so I was trying to create the above as a library so I could add odbc support, and then compare with the not-so-efficient `PostgreSQL binary <> Arrow` support that I have there.

> On another note. Personally I regret naming the tool odbc2parquet a little bit since I want to add functionality to also insert the content of parquet files into SQL Tables. I do lack a better name though. Since you might face kind of the same issue with odbc2arrow, I would be interested in your thoughts regarding naming, if you are up for some bike-shedding.

How about `arrodbc`? There are certainly convenience and performance benefits in having ODCB <> {CSV | Parquet | other} separate to Arrow, but perhaps Arrow can help keep the interface simpler, as it already supports reading and writing CSV and Parquet (I'd love to see us supporting more formats).
Maybe if all the `odbc2*` crates lived together as sub-crates of one repo, we could share a lot of the implementation with a core library, and wrap the binaries around it (especially once you have parameters and prepared statements, and more).

I'd love to work with you on these, I might not be helpful on the odbc side, but I should be useful on Arrow's side.
when you go through my implementation, you'll notice that I left a few TODOs for where I think we could create Arrow arrays quicker.

**Comment by pacman82** [comment]
Hi @nevi-me ,

there is lot's of things I want to write to you, but it is kind of late for me so I keep it short.

> So even if I take the cursor, I still end up with a lifetime problem.

I have some thoughts on the lifetime issues I'll be sharing with you then I've slept more. Also the `odbc2arrow` repository might be a better place for that.

> How about arrodbc?

Maybe we should not try to cute of a name and simply do `arrow-odbc` or `odbc-arrow`?


> I'd love to work with you on these, ...

I am looking forward to working with you on Arrow <-> ODBC as our expertises complement each other very nicely.

**Comment by nevi-me** [comment]
> I have some thoughts on the lifetime issues I'll be sharing with you then I've slept more. Also the `odbc2arrow` repository might be a better place for that.

Hi @pacman82, thanks. I'd appreciate that.

I'll have a look at what the Apache trademarks prescribe when it comes to names, but I think either `arrow-odbc` or `odbc-arrow` should suffice.

What do you think of the suggestion of then basing the various binaries around this common library layer?

> I am looking forward to working with you on Arrow <-> ODBC as our expertises complement each other very nicely.

I look forward to working with you too :)

**Comment by pacman82** [comment]
Hi @nevi-me 

> What do you think of the suggestion of then basing the various binaries around this common library layer?

I like the suggestion. It would also get rid of the overhead of maintaining the docker test setup in parallel for the various binaries. How would we handle versioning? A single version for everything? Still separate version numbers for the various binaries? How would we integrate this with GitHub Actions?

What would you think about making the odbc-arrow buffer a feature of `odbc-api` behind a compiler flag?

I have finished basic support for parameters and prepared queries. I am also currently working on making lifetime issues with regards to `Cursors` easier (or at least easier to understand for the user).

Independent of how we model this in the type systems there are two problems hiding beneath, the various compiler errors one might encounter during such an endavour:

1. The connection has to be open during data is fetched from the Database. This is pretty straight forward, and should be of little suprise to the users.
2. If the Cursor would take ownership of the bound buffer, the entire entity would become self referential. The reason for this is, that we pass pointers to the ODBC C Api containing the buffer positions. This is a problem since rusts allows for moving types in memory, if we move such a Cursor the C pointers COULD become invalid. At the moment `odbc-api` side steps the issue by borrowing the buffers for the time they are bound, preventing them from moving. However, I decideded to change the `RowSetBuffer` trait contract to demand that moving the type MUST NOT invalidate the pointers. This would make owning them safe. However the currently provided implementations would cause undefined behaviour, since they bind stack allocated members directly (and thus a move would render them invalid). Yet this is not to hard a change to implement. The new contract would also not prevent the user from still using stack allocated buffers, in such a case the contract would have to be implemented for the reference type.

To sum this up in a nutshell, I'll replace one of the explicit lifetime parameters with a lengthy comment in the `Safety` section of `RowSetBuffer`.

Regarding `Arrow` arrays. Once they are allocated, what actions could cause them to be reallocated?

**Comment by pacman82** [comment]
ping @alippai

**Comment by alippai** [comment]
Thanks @pacman82 for adding me. @nevi-me last time you discussed rust DB driver topic on the Arrow mailing list I tried to give some spotlight to turbodbc and odbc* projects created by Markus. Next time I'll be more direct, clear and pushy :P

I'm happy to test and participate in the development. While I don't have Rust experience, I recently wrote Vertica and PostgreSQL driver experiments in pure python (yes, it's slow, but better than the alternatives) and I'm an active Turbodbc user too.

**Comment by pacman82** [comment]
@alippai I'll greatly appreciate any help. Getting the usecases right is arguably they more important than any other aspect, so I am appreciative of any input you can give me. There is a lot todo besides writing Rust code. I started writing a guide/tutorial for first time users. Also I am improving the docker setup for local development, as the tests currently run on my machine, because I have the ODBC drivers installed, but as is we can not hope to attract many contributers. Once I have a reproducible local setup, I still want all of this to run in CI via GitHub Actions. So there is tons of stuff todo, and any help is greatly appreciated.

**Comment by nevi-me** [comment]
> @nevi-me last time you discussed rust DB driver topic on the Arrow mailing list I tried to give some spotlight to turbodbc and odbc* projects created by Markus. Next time I'll be more direct, clear and pushy :P

Hi @alippai @pacman82 , I wanted to defer my response to a time when I'd have something to show.

For Markus' benefit, the relevant mailing list thread is https://lists.apache.org/thread.html/r62eb07e36774909d0610db9dea98ec4c7235575cb4667afb6b8adc87%40%3Cdev.arrow.apache.org%3E

My idea is to create something a bit narrow, in that it is mainly focused on efficiently converting raw/binary data from a database into Arrow data, and the same when writing to a database. So it'd be similar to turbodbc's ability to fetch database records as Arrow batches.

The other part is an optimization extension on top of this narrow interface, effectively something that allows a database API to say "I know how to filter data, so tell me what you'd like to filter, and I'll write the SQL to do that".
This second part is a bit harder, as AFAIK we only have a SQL parser in the `sqlparser` crate, but we don't yet have anything in Rust that can convert an AST to SQL.

I've opened a draft PR for discussion in the Arrow project at https://github.com/apache/arrow/pull/8731, to start demonstrating how what I was thinking of, could look like.
I've plugged in a Postgres reader into DataFusion, but haven't yet gone beyond that.

We could theoretically add `odbc` support as a DB option, as we'd mainly need to be able to get an Arrow schema; then pass a SQL query and get Arrow results back.

Given Rust's typed nature, and the inability to reflect like JDBC in supporting drivers at runtime, it begs a question of whether the support for various DBs could/should live in the Arrow project, or whether the Arrow project should only expose traits, with consumers bringing their implementations. I'm still hazy on this, but I'll be experimenting more.

What I'd like to try out, is to add odbc as a source to my draft PR, so I can perform comparisons with the postgres source that I've already implemented. What are your thoughts?

**Comment by pacman82** [comment]
@nevi-me Just seeing this now. Thanks for including me into the conversation. I hope I find the time to take a closer look at it soon.

**Comment by alippai** [comment]
Finally I got time to look into both this repo and the [PR](https://github.com/apache/arrow/pull/8731) from @nevi-me.

@pacman82 I don't see any trivial optimizations if using ODBC - as you don't really have control over the connection, but see my question below. I find the new odbc-api really nice and the odbc2parquet really shows it's simplicity.

@nevi-me Speaking specifically about Postgres, it might be the `rust-postgres` limitation, but to get the schema you don't need to do any workarounds. When you call `execute` the first message from the server will be a `RowDescription` [https://www.postgresql.org/docs/9.3/protocol-flow.html#AEN99807](https://www.postgresql.org/docs/9.3/protocol-flow.html#AEN99807), then it's followed 0..N `DataRow`s. The `RowDescription` contains everything you'll need, just like it's available using `odbc-api`: [https://github.com/pacman82/odbc2parquet/blob/master/src/query.rs#L64](https://github.com/pacman82/odbc2parquet/blob/master/src/query.rs#L64). This can be found on the `tokio_parquet` crate's `Statement` module: [https://docs.rs/tokio-postgres/0.6.0/tokio_postgres/struct.Statement.html#method.columns](https://docs.rs/tokio-postgres/0.6.0/tokio_postgres/struct.Statement.html#method.columns)

I might be wrong, but one advantage of using the raw `PostgreSQL` wire protocol over ODBC makes it possible to add a fast-track for non-null, fixed length rows to be parsed in a vectorized way (e.g querying _a lot_ of numbers, IDs for data science). AFAIK while strided (mem)copy from an arrays is available in NEON instructions only (ARM), this special case could give you a slight improvement on x86 too. @pacman82 correct me, but even odbc2parquet can take (or already takes?) advantage of this if all columns are fixed width, non-nullable. 

@nevi-me I think the model you proposed is great in a hybrid environment eg. JOIN-ing a table in cold storage (parquet) with live data (postgresql), but also I'd say if you need to run a query more than once (I saw it for schema, limit/offset in the code) that would be out of scope for any generic database client. I don't think you want to deal with different isolation levels, long lived sessions, slow offset implementations of the DBs. One _"statement"_ per data-source should be a reasonable limit, even if you add streaming, fetching it batch by batch.

**Comment by alippai** [comment]
Also take my feedback with a grain of salt, please. I'm new at giving feedback or review on this topic or field, so if you don't understand something _it's me, not you_. I'm the 100% newbie and you two the professionals here :)

**Comment by pacman82** [comment]
Oh dear, sorry for taking so long to respond. I did have a nice christmas and a good start into the new year. I wish you guys a happy new year, too.

@alippai odbc2parquet could take advantage of these optimizations, in case of non null columns which are binary identical to their counterpart in the parquet abi, we could just pass the pointer through to the parquet writer without any copies at all. Might be helpful in some payloads. My guess (and it's just a guess, I do not have any benchmarks) is that most usecases would be restricted by IO, the remaining ones probaly be dominated be the allocations we have to perform writing the string columns. That doesn't mean that it wouldn't be a nead thing to do, though.

@nevi-me Considering the sheer number of databases / data sources, my gut feeling tells me it is probably better to offer support for brining your on data source in the arrow projects, rather than to try to include them all into the arrow project. Withr regards to ODBC, at one point I'd like to explore the approach you took with odbc2arrow further, and see how it works out.

**Comment by pacman82** [comment]
> I might be wrong, but one advantage of using the raw `PostgreSQL` wire protocol over ODBC makes it possible to add a fast-track for non-null, fixed length rows to be parsed in a vectorized way (e.g querying _a lot_ of numbers, IDs for data science). AFAIK while strided (mem)copy from an arrays is available in NEON instructions only (ARM), this special case could give you a slight improvement on x86 too. @pacman82 correct me, but even odbc2parquet can take (or already takes?) advantage of this if all columns are fixed width, non-nullable. (edit: I can imagine that such optimization would require JIT, so this might not be a useful idea)

@alippai So I have no benchmarks and do not know about instruction level. Yet at the very least I can say that starting in version `0.6.11` `odbc2parquet` puts the memory filled by ODBC as is into parquet library without any further processing into the `write_batch` method of the `parquet` crate.

I wanted to do this optimization, back then I started the artefact, but I only got to it just now. Since you are very performance minded I thought you might want to know.

See: https://github.com/pacman82/odbc2parquet/commit/6a372e2f2f904667d969c6d4afaf19a6cde42f0d

**Comment by pacman82** [comment]
@nevi-me 

This idea has been a long time in the making. I feel `odbc-api` is now mature enough to start with `odbc-arrow`. I started work on it in this repository: <https://github.com/pacman82/odbc-arrow>. I would not mind if an experienced Arrow would look over it.

Cheers, Markus

**Comment by pacman82** [comment]
Closing this issue due to the release of `arrow-odbc`.
